### PR TITLE
feat: C509 `Name` type

### LINKED ISSUE #589: 🛠️ [TASK] : Implement encode decode for C509 type `Name`
### Summary

 Implement encode decode for C509 type `Name`

### Description

 Implement encode, decode, and other important functionality for C509 type `Name`

Given
```
    Name = [ * RelativeDistinguishedName ] / text / bytes

    RelativeDistinguishedName = Attribute / [ 2* Attribute ]

    Attribute = ( attributeType: int, attributeValue: text ) //
                ( attributeType: ~oid, attributeValue: bytes ) //
                ( attributeType: pen, attributeValue: bytes )
```
This type `Name` will be used in various places like `GeneralName`, `subject`, and `issuer`

Note that `Attribute` is used in various places

```
      Attributes = ( attributeType: int, attributeValue: [+text] ) //
                   ( attributeType: ~oid, attributeValue: [+bytes] )
      SubjectDirectoryAttributes = [+Attributes]
```

`Attribute` used in `RelativeDistinguishedName` support PEN
`Attribute` used in `Attributes` doesn't support PEN and support multiple `Attribute` value

The implementation should follow the c509 documentation https://datatracker.ietf.org/doc/draft-ietf-cose-cbor-encoded-cert/09/

- [x] Create an `Attribute` type decode and encode where it will handle attributeType integer value, unwrapped OID, and PEN
- [x] Create a `RelativeDistinguishedName` type decode and encode
- [x] Create a `Name` type decode and encode
- [x] Modify the `GeneralName` to accept `Name` type
- [x] All special cases should be handled
- [x] Create tests for all use cases

Test cases

**Attribute**
- `Attribute` should work for both `RelativeDistinguishedName` and `Attributes`.
- `Attribute` should not have empty `attributeValue`
- `Attributes` shoudl not be empty array
- Each Attribute has corresponding integer value associated with the OID (Table 9.3)

**Relative Distinguised Name (RDN)**
- If only > 1 attributes encode it as array

**Name**
- 3 special cases for `Name` `CommonName`
- `Name` can contains >= 0 `RelativeDistinguisedName` if the`RelativeDistinguisedName` is provided

### PR DESCRIPTION
**Pull request by bkioshn**
# Description

 Implement encode and decode for C509 `Name` type

## Related Issue(s)

#589 

## Description of Changes
- Implement encode and decode for type `Name`, `RelativeDistinguisedName`, `Attribute`, `Attributes`
- Handle special cases
- Change `DirectoryName` type from `Unsupported` to `Name` in `GeneralName`
- Test cases added 

## Please confirm the following checks

* [x] My code follows the style guidelines of this project
* [x] I have performed a self-review of my code
* [x] I have commented my code, particularly in hard-to-understand areas
* [x] I have made corresponding changes to the documentation
* [x] My changes generate no new warnings
* [x] I have added tests that prove my fix is effective or that my feature works
* [x] New and existing unit tests pass locally with my changes
* [x] Any dependent changes have been merged and published in downstream module

### CODE REVIEW

File: catalyst-gateway-crates/c509-certificate/src/c509_attributes/data.rs [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+
+    for data in ATTRIBUTE_DATA {
+        int_to_oid_table.add(data.0, data.1);
+    }
+
+    AttributeData { int_to_oid_table }
+});
+
+/// Static reference to the `AttributeData` lookup table.
+pub(crate) static ATTRIBUTES_LOOKUP: &Lazy<AttributeData> = &ATTRIBUTES_TABLES;
```
Comment (stevenj) [inline catalyst-gateway-crates/c509-certificate/src/c509_attributes/data.rs:76]: Don't have `pub(crate)` on static structures.
Make it private to the file and accessor functions to query/update it.

**Review by stevenj — CHANGES_REQUESTED** [review]
Looks good except for my single request for change.

File: catalyst-gateway-crates/c509-certificate/src/c509_attributes/data.rs (hunk 1 above)
Comment (stevenj) [reply catalyst-gateway-crates/c509-certificate/src/c509_attributes/data.rs:76]: I am ok with this, now, given how it is used.

**Review by stevenj — APPROVED** [review]
LGTM

**Review by apskhem — APPROVED** [review]

**Pull request merged**